### ISSUE TITLE
learning_rate is too large in Multi-Layer Perception example

### ISSUE THREAD

**Issue by berryfl** [ml-explore/mlx#188]
In the documentation here, the learning_rate is set to 1e-1, which seems too large to get meaningful results.

In my Macbook Pro, I always get the accuracy value 0.113

```
Epoch 0: Test accuracy 0.113
Epoch 1: Test accuracy 0.113
Epoch 2: Test accuracy 0.113
Epoch 3: Test accuracy 0.113
Epoch 4: Test accuracy 0.113
Epoch 5: Test accuracy 0.113
Epoch 6: Test accuracy 0.113
Epoch 7: Test accuracy 0.113
Epoch 8: Test accuracy 0.113
Epoch 9: Test accuracy 0.113
```

After I change the learning_rate to 1e-3 and epochs to 20, the results seems much better

```
Epoch 0: Test accuracy 0.851
Epoch 1: Test accuracy 0.887
Epoch 2: Test accuracy 0.902
Epoch 3: Test accuracy 0.914
Epoch 4: Test accuracy 0.919
Epoch 5: Test accuracy 0.923
Epoch 6: Test accuracy 0.924
Epoch 7: Test accuracy 0.932
Epoch 8: Test accuracy 0.933
Epoch 9: Test accuracy 0.933
Epoch 10: Test accuracy 0.937
Epoch 11: Test accuracy 0.940
Epoch 12: Test accuracy 0.938
Epoch 13: Test accuracy 0.942
Epoch 14: Test accuracy 0.943
Epoch 15: Test accuracy 0.943
Epoch 16: Test accuracy 0.946
Epoch 17: Test accuracy 0.945
Epoch 18: Test accuracy 0.946
Epoch 19: Test accuracy 0.947
```

**Comment by awni** [comment]
Can you share the code you are running? I just ran the example in [mlx-examples](https://github.com/ml-explore/mlx-examples/tree/main/mnist) and it gives pretty reasonable accuracy:

```
Epoch 0: Test accuracy 0.866, Time 0.092 (s)
Epoch 1: Test accuracy 0.896, Time 0.089 (s)
Epoch 2: Test accuracy 0.917, Time 0.091 (s)
Epoch 3: Test accuracy 0.928, Time 0.088 (s)
Epoch 4: Test accuracy 0.935, Time 0.088 (s)
Epoch 5: Test accuracy 0.938, Time 0.088 (s)
Epoch 6: Test accuracy 0.946, Time 0.087 (s)
Epoch 7: Test accuracy 0.941, Time 0.088 (s)
Epoch 8: Test accuracy 0.950, Time 0.088 (s)
Epoch 9: Test accuracy 0.948, Time 0.088 (s)
```

**Comment by berryfl** [comment]
Sure, I didn't realize there is `mlx-examples` repo there, so I just input the code

```main.py
import mlx.core as mx
import mlx.nn as nn
import mlx.optimizers as optim

import numpy as np

class MLP(nn.Module):
    def __init__(
        self,
        num_layers: int,
        input_dim: int,
        hidden_dim: int,
        output_dim: int,
    ):
        super().__init__()
        layer_sizes = [input_dim] + [hidden_dim] * num_layers + [output_dim]
        self.layers = [
            nn.Linear(idim, odim)
            for idim, odim in zip(layer_sizes[:-1], layer_sizes[1:])
        ]

    def __call__(self, x):
        for l in self.layers[:-1]:
            x = mx.maximum(l(x), 0.0)
        return self.layers[-1](x)

def loss_fn(model, X, y):
    return mx.mean(nn.losses.cross_entropy(model(X), y))

def eval_fn(model, X, y):
    return mx.mean(mx.argmax(model(X), axis=1) == y)

num_layers = 2
hidden_dim = 32
num_classes = 10
batch_size = 256
num_epochs = 10
learning_rate = 1e-3

# Load the data
import mnist
train_images, train_labels, test_images, test_labels = map(
    mx.array,
    (
        mnist.get_train_images(),
        mnist.get_train_labels(),
        mnist.get_test_images(),
        mnist.get_test_labels(),
    ),
)

def batch_iterate(batch_size, X, y):
    perm = mx.array(np.random.permutation(y.size))
    for s in range(0, y.size, batch_size):
        ids = perm[s : s + batch_size]
        yield X[ids], y[ids]

# Load the model
model = MLP(num_layers, train_images.shape[-1], hidden_dim, num_classes)
mx.eval(model.parameters())

# Get a function which gives the loss and gradient of the
# loss with respect to the model's trainable parameters
loss_and_grad_fn = nn.value_and_grad(model, loss_fn)

# Instantiate the optimizer
optimizer = optim.SGD(learning_rate=learning_rate)

for e in range(num_epochs):
    for X, y in batch_iterate(batch_size, train_images, train_labels):
        loss, grads = loss_and_grad_fn(model, X, y)

        # Update the optimizer state and model parameters
        # in a single call
        optimizer.update(model, grads)

        # Force a graph evaluation
        mx.eval(model.parameters(), optimizer.state)

    accuracy = eval_fn(model, test_images, test_labels)
    print(f"Epoch {e}: Test accuracy {accuracy.item():.3f}")
    ```
    
    ```mnist.py
    import array
import gzip

def get_images(filename):
    f = gzip.open(filename)
    magic_number = int.from_bytes(f.read(4), "big") # Should be 2051
    num_items = int.from_bytes(f.read(4), "big")
    num_rows = int.from_bytes(f.read(4), "big")
    num_cols = int.from_bytes(f.read(4), "big")
    images = []
    for _ in range(num_items):
        image = array.array('B', f.read(num_rows * num_cols))
        images.append(image.tolist())
    f.close()
    return images

def get_labels(filename, label_as_tuple=False):
    f = gzip.open(filename)
    magic_number = int.from_bytes(f.read(4), "big") # Should be 2049
    num_items = int.from_bytes(f.read(4), "big")
    labels = []
    for _ in range(num_items):
        label = int.from_bytes(f.read(1), "big")
        if label_as_tuple:
            label = tuple(((1 if i == label else 0) for i in range(10)))
        labels.append(label)
    f.close()
    return labels

def get_train_images():
    return get_images("train-images-idx3-ubyte.gz")

def get_train_labels():
    return get_labels("train-labels-idx1-ubyte.gz")

def get_test_images():
    return get_images("t10k-images-idx3-ubyte.gz")

def get_test_labels():
    return get_labels("t10k-labels-idx1-ubyte.gz")
```

**Comment by berryfl** [comment]
looks like when I use the mnist.py in the example, I can get similar results now.
